### PR TITLE
added support of b2b bmm

### PR DESCRIPTION
**Pull request by PivovarA**
# Summary

Updated CUTLASS example back-to-back gemm to back-to-back batch gemm.
I've added b2b bmm support to the kernel/b2b_gemm.h and modified parameters and arguments in b2b_gemm.h in device and kernel.

# Test Plan

> make 13_two_tensor_op_fusion
> ./examples/13_two_tensor_op_fusion/13_fused_two_gemms_s8_sm80_rf

```
Device: NVIDIA PG509-210
Arch: SM80
Test: gemm int8 RF residency
Running Non-fused back-to-back INT8 NT interleaved GEMMs...
gemm 0 time 0.0553379 ms
gemm 1 time 0.0265773 ms
Non-fusion time 0.0819152 ms
Pass
Running Fused back-to-back INT8 NT interleaved GEMMs with RF residency...
Fusion time 0.0594058 ms
Pass
Running Fused back-to-back INT8 NT interleaved Batched GEMMs with RF residency...
Fusion time 0.00709632 ms
Pass
```

### CODE REVIEW

File: examples/13_two_tensor_op_fusion/device/b2b_gemm.h [hunk 1]
```diff
@@ -295,12 +313,12 @@ class B2bGemm {
...
-      
+
     // Determine grid shape
     ThreadblockSwizzle threadblock_swizzle;
 
     cutlass::gemm::GemmCoord tiled_shape = threadblock_swizzle.get_tiled_shape(
-      args.problem_size_0, 
+      args.problem_size_0,
       {ThreadblockShape0::kM, ThreadblockShape0::kN, ThreadblockShape0::kK},
       args.split_k_slices);
```
Comment (ipiszy) [inline examples/13_two_tensor_op_fusion/device/b2b_gemm.h:323]: We could use batch_count to initialize this, refer to https://github.com/NVIDIA/cutlass/blob/a31b43b3f3cb609c6ed0b30ca330d766291e84fb/include/cutlass/gemm/kernel/params_universal_base.h#L144.

File: examples/13_two_tensor_op_fusion/device/b2b_gemm.h [hunk 2]
```diff
@@ -233,12 +240,18 @@ class B2bGemm {
...
+      int64_t batch_stride_B0 = 1,
+      int64_t batch_stride_B1 = 1,
+      int64_t batch_stride_C1 = 1,
+      int64_t batch_stride_D1 = 1,
+      typename EpilogueOutputOp0::Params epilogue0_ =
         typename EpilogueOutputOp0::Params(),
-      typename EpilogueOutputOp1::Params epilogue1_ = 
+      typename EpilogueOutputOp1::Params epilogue1_ =
         typename EpilogueOutputOp1::Params(),
       int split_k_slices_ = 1
```
Comment (ipiszy) [inline examples/13_two_tensor_op_fusion/device/b2b_gemm.h:252]: Please add a check to make sure that gemm_batched and split_k_slices are not used at the same time.

File: examples/13_two_tensor_op_fusion/device/b2b_gemm.h [hunk 3]
```diff
@@ -196,6 +196,7 @@ class B2bGemm {
     // Data members
     //
 
+    GemmUniversalMode mode;
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply examples/13_two_tensor_op_fusion/device/b2b_gemm.h:199]: Please add this argument and update test cases accordingly.

**Review by ipiszy — COMMENTED** [review]
Thanks @apivovarov!

**Comment by apivovarov** [comment]
> Thanks @apivovarov!

@ipiszy, this PR was created by @PivovarA

**Comment by ipiszy** [comment]
> > Thanks @apivovarov!
> 
> @ipiszy, this PR was created by @PivovarA

Ah sorry my bad..

File: examples/13_two_tensor_op_fusion/fused_two_gemms_s8_sm80_rf.cu [hunk 4]
```diff
@@ -201,7 +201,15 @@ bool run_fused_gemm_s8_sm80_rf_res() {
   B2bInterleavedFusedGemmRun<B2bGemm, 32> fusedGemm;
 
   std::cout << "Running Fused back-to-back INT8 NT interleaved GEMMs with RF residency...\n";
-  bool passed = fusedGemm.run(gemm_s8_sm80_problem_size_0, gemm_s8_sm80_problem_size_1, alpha0, beta0, alpha1, beta1);
+  bool passed = fusedGemm.run(
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply examples/13_two_tensor_op_fusion/fused_two_gemms_s8_sm80_rf.cu:204]: You could replace Gemm with BatchedGemm for reference value calculation in b2b_gemm_run.h. https://github.com/NVIDIA/cutlass/blob/a31b43b3f3cb609c6ed0b30ca330d766291e84fb/tools/util/include/cutlass/util/reference/device/gemm.h#L308

**Review by ipiszy — COMMENTED** [review]
Thanks [PivovarA](https://github.com/PivovarA)! looks good overall, please remove unnecessary SplitKSerial usage and add batch_size>1 test cases.

**Review by ipiszy — COMMENTED** [review]
Thanks @PivovarA! Test cases need to be refined.

**Comment by mnicely** [comment]
Hi @PivovarA , thanks for the PR! Does this only support INT8?

**Comment by ipiszy** [comment]
> Hi @PivovarA , thanks for the PR! Does this only support INT8?

This change should be orthogonal to data types.

**Comment by ipiszy** [comment]
Thanks @PivovarA ! lgtm~

File: examples/13_two_tensor_op_fusion/device/b2b_gemm.h [hunk 5]
```diff
@@ -295,16 +311,16 @@ class B2bGemm {
...
+      args.problem_size_0,
       {ThreadblockShape0::kM, ThreadblockShape0::kN, ThreadblockShape0::kK},
-      args.split_k_slices);
+      args.batch_count);
 
-    if (kSplitKSerial && args.split_k_slices > 1) {
+    if (args.mode == GemmUniversalMode::kGemm && args.batch_count > 1) {
 
 
       bytes += sizeof(int) * size_t(tiled_shape.m()) * size_t(tiled_shape.n());
```
[the remark this answers is not in the document]
Comment (hwu36) [reply examples/13_two_tensor_op_fusion/device/b2b_gemm.h:326]: it is used to store the semaphore for splitk.

File: examples/13_two_tensor_op_fusion/device/b2b_gemm.h [hunk 6]
```diff
@@ -320,38 +336,33 @@ class B2bGemm {
...
+      args.batch_count);
 //    cutlass::gemm::GemmCoord grid_shape_1 = threadblock_swizzle.get_tiled_shape(
-//      args.problem_size_1, 
+//      args.problem_size_1,
 //      {ThreadblockShape1::kM, ThreadblockShape1::kN, ThreadblockShape1::kK},
 //      args.split_k_slices);
 
-    if (kSplitKSerial) {
-      if (args.split_k_slices > 1) {
+    if (args.mode == GemmUniversalMode::kGemm) {
```
Comment (hwu36) [inline examples/13_two_tensor_op_fusion/device/b2b_gemm.h:347]: we can remove this if .. else .. part too

File: examples/13_two_tensor_op_fusion/device/b2b_gemm.h [hunk 7]
```diff
@@ -373,8 +391,8 @@ class B2bGemm {
 
   /// Lightweight update given a subset of arguments
   Status update(Arguments const &args, void *workspace = nullptr) {
-    
-    if (kSplitKSerial && args.split_k_slices > 1) {  
+
+    if (args.mode == GemmUniversalMode::kGemm && args.batch_count > 1) {
```
Comment (hwu36) [inline examples/13_two_tensor_op_fusion/device/b2b_gemm.h:395]: this one is not needed

**Review by hwu36 — APPROVED** [review]

**Pull request merged**